File: python/heap/find_k_pair.py
import heapq

from typing import List, Tuple

class Pair:
    def __init__(self, num1: int, num2: int, i: int, j: int):
        self.num1 = num1
        self.num2 = num2
        self.i = i
        self.j = j

    def __lt__(self, other: 'Pair') -> bool:
        return (self.num1 + self.num2) < (other.num1 + other.num2)


class Solution:
    def k_smallest_pairs(self, nums1: List[int], nums2: List[int], k:int) -> List[List[int]]:
        if not nums1 or not nums2 or k == 0:
            return []

        min_heap: List[Pair] = []
        result: List[List[int]] = []

        for i in range(min(k, len(nums1))):
            heapq.heappush(min_heap, Pair(nums1[i], nums2[0], i, 0))

        while k > 0 and min_heap:
            pair = heapq.heappop(min_heap)

            result.append([pair.num1, pair.num2])
            k -= 1

            if pair.j + 1 < len(nums2):
                next_pair = Pair(
                    pair.num1,
                    nums2[pair.j +1],
                    pair.i,
                    pair.j + 1
                )

                heapq.heappush(min_heap, next_pair)


        return result

File: golang/heap/find_k_pair.go
package main

import "container/heap"

type Pair struct {
	num1, num2, i, j int
}

type PriorityQueue []Pair

func (pq PriorityQueue) Len() int {
	return len(pq)
}

func (pq PriorityQueue) Less(i, j int) bool {
	return (pq[i].num1 + pq[i].num2) < (pq[j].num1 + pq[j].num2)
}
func (pq PriorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(Pair))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)

	x := old[n-1]

	*pq = old[0 : n-1]

	return x
}

func kSmallestPairs(nums1 []int, nums2 []int, k int) [][]int {
	if len(nums1) == 0 || len(nums2) == 0 || k == 0 {
		return [][]int{}
	}

	pq := &PriorityQueue{}

	heap.Init(pq)

	for i := 0; i < len(nums1) && i < k; i++ {
		heap.Push(pq, Pair{nums1[i], nums2[0], i, 0})
	}

	result := [][]int{}

	for k > 0 && pq.Len() > 0 {
		pair := heap.Pop(pq).(Pair)

		result = append(result, []int{pair.num1, pair.num2})

		k--

		if pair.j+1 < len(nums2) {
			heap.Push(pq, Pair{pair.num1, nums2[pair.j+1], pair.i, pair.j + 1})
		}
	}

	return result
}
